### PR TITLE
Add notion of fingerprint type.

### PR DESCRIPTION
**Pull request by ndmckinley**
Fingerprint type: automatically exclude from generation and mark as output.
Make disk use this rather than custom code in terraform.
Enables future use of fingerprints on other resources without custom code.

-----------------------------------------------------------------
# [all]
No changes expected
## [terraform]
Minor cleanup of disk's labelFingerprint.

### CODE REVIEW

File: api/type.rb [hunk 1]
```diff
@@ -192,6 +192,19 @@ def initialize(name)
       NAME = Api::Type::String.new('name')
     end
 
+    # Represents a fingerprint.  A fingerprint is an output-only
+    # field used for optimistic locking during updates.
+    class Fingerprint < String
+      def validate
+        super
+        @output = true if @output.nil?
```
Comment (rambleraptor) [inline api/type.rb]: I think you could do `@output ||= true` here.

File: api/type.rb [hunk 2]
```diff
@@ -192,6 +192,19 @@ def initialize(name)
...
+        # Fingerprints are not usually worth including in the
+        # list of output fields.  This should be overridden in
+        # Terraform, where outputting fingerprints is idiomatic.
+        @exclude ||= true
```
Comment (rambleraptor) [inline api/type.rb]: The comment confuses me a little bit. It seems like the comment goes along with line 200.

As for making `exclude = true` the default, that's temporary, right? Could you put a TODO or something to mark that as temporary?

File: api/type.rb (hunk 1 above)
Comment (ndmckinley) [reply api/type.rb]: Sorry, that'll behave wrong if output is set to false.

```ruby
irb(main):008:0> a = false
=> false
irb(main):009:0> a ||= true
=> true
irb(main):010:0> a
=> true
```

File: api/type.rb (hunk 2 above)
Comment (ndmckinley) [reply api/type.rb]: I suppose it is temporary!  We have to work out how fingerprints ought to behave in puppet, chef, ansible, which means we need custom updates in those.  Added the TODO.

**Comment by rambleraptor** [comment]
I’m on mobile, but that looks awesome.

(Side note - I picked up the habit of being sometimes pedantic with TODOs.
Let me know if it seems unnecessary)

**Comment by ndmckinley** [comment]
I'm treating that as an LGTM and merging.  :)  You were totally right to comment on that, thanks for doing so!

**Pull request merged**